### ISSUE TITLE
Upgrade Issue from 1.8.3 to 1.9.0

### ISSUE THREAD

**Issue by boombuler** [linq2db/linq2db#837]
We have a custom IDateContext implementation, because we need to run queries in the same transaction as our legacy code and also to use our own DbConnection implementation. 

We just tried to upgrade to 1.9.0 but due to the changes to IDataContext we can't do this anymore because IDataContextEx is not accessable. 

We also get some `InvalidCastException`s because of the typecasts in `ExpressionQuery<T>` and some other locations. For example:
```
Exception message: Unable to cast object of type '****.DataContext' to type 'LinqToDB.IDataContextEx'
Stack trace:
at LinqToDB.Linq.QueryRunner.GetSqlText(Query query, IDataContext dataContext, Expression expr, Object[] parameters, Int32 idx)
   at LinqToDB.Linq.ExpressionQuery`1.get_SqlText()
   -- snip --
```

### Environment details
linq2db version: *1.9.0*  
Framework version: *.NET Core 2*

**Comment by sdanyliv** [comment]
Could you please share your implemenattion somewhere or sample usage, maybe we can create solution for next release, that planned on this week.

**Comment by boombuler** [comment]
The implementation is mostly equal to the DataContext implementation of 1.8.3 but it uses our own DbConnection implementation to run the queries.
So I have a different implementation for those parts which are now moved to the IQueryRunner interface. 
These are modified to share the same database configuration and the same transactions as our legacy code.

I cannot share the implementation itself because it would require to publish some larger parts of a closed source software.

I think setting IDataContextEx and IQueryRunner to public would solve my issues or maybe move the IDataContextEx back to IDataContext because it seems to me that nearly all parts of linq2db require IDataContext implementations to also implement IDataContextEx.

**Comment by sdanyliv** [comment]
I can make these interfaces public if there is no obligations from the team. 
But for clarification, why you do not use DataConnection with .ctor that receives transaction as parameter?

```cs
/// <summary>
/// Creates database connection object that uses specified database provider and transaction.
/// </summary>
/// <param name="dataProvider">Database provider implementation to use with this connection.</param>
/// <param name="transaction">Existing database transaction to use.</param>
public DataConnection(
	[JetBrains.Annotations.NotNull] IDataProvider dataProvider,
	[JetBrains.Annotations.NotNull] IDbTransaction transaction)
```

**Comment by ili** [comment]
Hi! @boombuler 

Can you please make PR with changes, that will make your `IDataContext` implementation possible

**Comment by boombuler** [comment]
Hi,

I've created a PR #849.
I think it's a good idea to remove the new IDataContextEx interface since it completely removes the possibility for external IDataContext implementations.

If you have public interfaces, be aware that people might implement them :) 

@sdanyliv We didn't use this constructor because we have some more changes like a MappingSchema that reads the metadata from the database and we also have some code that rewrites the query before it is executed depending on the database provider. And it finally results in IQueryable<MyTable> is injectable with a DI-Container... So the DataContext  will not be directly constructed and the existance of an transaction might change within a codeblock which doesn't know anything about database transactions.

**Comment by ili** [comment]
Great! I think we'll merge it after @igor-tkachev's review

**Comment by sdanyliv** [comment]
Fixed in  #849
